flake8_trio.py

```diff
@@ -58,6 +69,36 @@ def check_for_trio100(self, node: Union[ast.With, ast.AsyncWith]) -> None:
                     make_error(TRIO100, item.lineno, item.col_offset, call)
                 )
 
+    def trio101_mark_yields_safe(
+        self, node: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+    ) -> None:
+        if any(
+            isinstance(d, ast.Name)
+            and d.id in ("contextmanager", "asynccontextmanager")
+            for d in node.decorator_list
+        ):
+            self.safe_yields.update(
+                {x for x in ast.walk(node) if isinstance(x, ast.Yield)}
+            )
+
+    def check_for_trio101(self, node: Union[ast.With, ast.AsyncWith]) -> None:
+        for item in (i.context_expr for i in node.items):
+            call = is_trio_call(
+                item,
+                "open_nursery",
+                "fail_after",
+                "fail_at",
+                "move_on_after",
+                "move_at",
+            )
+            if call and any(
+                isinstance(x, ast.Yield) and x not in self.safe_yields
+                for x in ast.walk(node)
```

Comment:
Instead of collecting and marking all the yields, how about we track a `self._yield_is_error` boolean?  Then:

```python
    def visit_AsyncWith(self, node: ast.AsyncWith) -> None:
        self.check_for_trio100(node)
        outer = self._yield_is_error
        self._yield_is_error = any(is_trio_call(i.context_expr, "open_nursery") for i in node.items)
        self.generic_visit(node)
        self._yield_is_error = outer

    # similarly in visit_With for the sync context managers

    # visit_(Async)FunctionDef can set it to False, do generic_visit, set back to outer

    def visit_Yield(self, node: ast.Yield) -> None:
        if self._yield_is_error:
            self.problems.append(...)  # we want to point this error at the yield statement, not the cxt manager
        self.generic_visit(node)
```

Should be a bit cheaper since we avoid the extra iteration, and it's a generally useful pattern which I'd expect to come up again.